### ISSUE TITLE
General train issues

### ISSUE THREAD

**Issue by yannbcf** [citizenfx/fivem#2197]
I've been digging around to implement trains in the fiveM ecosystem and I faced few issues.

## 1st approach

My first approach was to create the trains server side via the CreateVehicleServerSetter method, here is some pseudo code
> server
```lua
RegisterCommand("train", function(source)
    local train = CreateVehicleServerSetter(`metrotrain`, 'train', -888.1231, -2310.7957, -12.6248, 0.0)
    print(GetVehicleType(train)) -- 'train'
    Wait(50)

    TriggerClientEvent("test", source, NetworkGetNetworkIdFromEntity(train))
end, false)
```

> client

The train renders as derailed
```lua
RegisterNetEvent("test", function(trainNetId)
    local fwTrain = NetworkGetEntityFromNetworkId(train)
    NetworkRequestControlOfEntity(fwTrain)
    Wait(500)

    SetMissionTrainCoords(fwTrain, -891.9408, -2321.6558, -12.6177) -- works
    SetRenderTrainAsDerailed(fwTrain, false) -- no effect
    SetTrainSpeed(fwTrain, 1.0) -- fails to execute the native
end)
```

## 2nd approach

My other idea was then to implement a serverside train streamer to create my own sync layer with clientside entities, and simulate the trains movement by interpolating the trains sector nodes when no player stream that virtual entity, this way the system would work even in relaxed lockdown mode, by the way, having the NetworkTime exposed server side would make such time-based sync really easy

> client
```lua
    loadTrainModels()

    local fwTrain = CreateMissionTrain(26, pos.x, pos.y, pos.z, direction, false, true) -- the 6th argument supposed to be 'isNetwork'
    print(NetworkGetEntityIsNetworked(fwTrain)) -- true

    freeTrainModels()
```

This makes a networked train, that is blocked by the relaxed lockdown mode, which is really meh

## Solutions

Well, I see few solutions

- Make the server-side created trains entities work with the client-side mission train natives
and / or
- Make the client-side native CreateMissionEntity 6th 'isNetwork' param work
and / or
- Implement a server side event regarding the lockdown modes, where we can accept or deny an entity created by a client, therefore we would have the full control, and it's nice / secured

**Comment by BerkieBb** [comment]
> Implement a server side event regarding the lockdown modes, where we can accept or deny an entity created by a client, therefore we would have the full control, and it's nice / secured

I think you're already able to do this with the server side [entityCreating](https://docs.fivem.net/docs/scripting-reference/events/server-events/#entitycreating) event

**Comment by yannbcf** [comment]
Yes, but if the bucket 0 is already in relaxed mode for example, you're not able to override the deletion mechanism, from this event by accepting the entity, I think

This event seem to only be useful when you're making a gamemode where you dont use the lockdown mode at all and decide which entity is legit

**Comment by Azerothwav** [comment]
Hi, i just find a fix for create trains networked with a lockdown activate, i will include it when i can to my script az_train (the solution is really dumb)

**Comment by gottfriedleibniz** [comment]
> ... create the trains server side via the CreateVehicleServerSetter method,

This will not work due to `CTrainGameStateData` not being constructed server-side.

While it is possible to create this server side. Without knowledge of `train_tracks/train_configs` there will be issues w/ the track and config fields that will make the API difficult to work with (esp. across game builds).

> Well, I see few solutions

To supplement. Another, albeit temporary, solution could be to expose CreateMissionTrain via RPC w/ server-side equivalents for the other mission train natives. This would cover a few things reported here (and requests elsewhere).

In both cases, the server-back-to-client reassignment may have some quirks due to the track node (+ relative position) not being explicitly synchronized (corrected on blender update). The RPC approach will also require its own changes/protocols for carriage creationTokens and etc.
